File: neuroshell.py
#!/usr/bin/env python3
"""
NeuroShell LKM  Python Interface
A utility for querying system hardware information via NeuroShell kernel module
"""

import os
import json
import sys
from pathlib import Path
from typing import Dict, Optional, Any

NEUROSHELL_PATH = Path("/sys/kernel/neuroshell")

class NeuroShellError(Exception):
    """Exception raised when NeuroShell module is not available"""
    pass

class NeuroShell:
    """Interface to NeuroShell kernel module"""
    
    def __init__(self):
        """Initialize NeuroShell interface"""
        if not NEUROSHELL_PATH.exists():
            raise NeuroShellError(
                "NeuroShell module not loaded. "
                "Please run 'sudo insmod neuroshell_enhanced.ko' first."
            )
    
    def read_attr(self, attribute: str) -> Optional[str]:
        """
        Read a NeuroShell sysfs attribute
        
        Args:
            attribute: Name of the attribute file
            
        Returns:
            Content of the attribute as string, or None if not found
        """
        attr_path = NEUROSHELL_PATH / attribute
        try:
            return attr_path.read_text().strip()
        except FileNotFoundError:
            return None
        except PermissionError:
            print(f"Permission denied reading {attribute}", file=sys.stderr)
            return None
    
    def parse_key_value(self, content: str) -> Dict[str, Any]:
        """
        Parse key=value format content
        
        Args:
            content: String with key=value pairs
            
        Returns:
            Dictionary of parsed values
        """
        result = {}
        for line in content.split('\n'):
            line = line.strip()
            if '=' in line:
                key, value = line.split('=', 1)
                # Try to convert to int
                try:
                    result[key] = int(value)
                except ValueError:
                    result[key] = value
        return result
    
    def get_cpu_info(self) -> Dict[str, Any]:
        """Get CPU information"""
        info = {}
        
        # Basic counts
        cpu_count = self.read_attr('cpu_count')
        if cpu_count:
            info['online'] = int(cpu_count)
        
        cpu_total = self.read_attr('cpu_total')
        if cpu_total:
            info['total'] = int(cpu_total)
        
        # Detailed info
        cpu_info = self.read_attr('cpu_info')
        if cpu_info:
            info.update(self.parse_key_value(cpu_info))
        
        # Topology
        topology = self.read_attr('cpu_topology')
        if topology:
            info['topology'] = topology
        
        return info
    
    def get_memory_info(self) -> Dict[str, Any]:
        """Get memory information"""
        info = {}
        
        # Total bytes
        mem_total = self.read_attr('mem_total_bytes')
        if mem_total:
            total_bytes = int(mem_total)
            info['total_bytes'] = total_bytes
            info['total_mb'] = total_bytes // (1024 * 1024)
            info['total_gb'] = total_bytes / (1024 * 1024 * 1024)
        
        # Detailed info
        mem_info = self.read_attr('mem_info')
        if mem_info:
            info.update(self.parse_key_value(mem_info))
        
        return info
    
    def get_numa_info(self) -> Dict[str, Any]:
        """Get NUMA information"""
        info = {}
        
        # Node count
        numa_nodes = self.read_attr('numa_nodes')
        if numa_nodes:
            info['nodes'] = int(numa_nodes)
        
        # Detailed info
        numa_info = self.read_attr('numa_info')
        if numa_info:
            info['details'] = numa_info
        
        return info
    
    def get_gpu_info(self) -> Dict[str, Any]:
        """Get GPU information"""
        info = {}
        
        # Summary
        gpu_info = self.read_attr('gpu_info')
        if gpu_info:
            info.update(self.parse_key_value(gpu_info))
        
        # Details
        gpu_details = self.read_attr('gpu_details')
        if gpu_details:
            info['details'] = gpu_details
        
        return info
    
    def get_accelerator_info(self) -> Dict[str, Any]:
        """Get AI accelerator information"""
        info = {}
        
        # Count
        accel_count = self.read_attr('accelerator_count')
        if accel_count:
            info['count'] = int(accel_count)
        
        # Details
        accel_details = self.read_attr('accelerator_details')
        if accel_details:
            info['details'] = accel_details
        
        return info
    
    def get_system_summary(self) -> str:
        """Get complete system summary"""
        return self.read_attr('system_summary') or "Summary not available"
    
    def get_all_info(self) -> Dict[str, Any]:
        """Get all available information"""
        return {
            'cpu': self.get_cpu_info(),
            'memory': self.get_memory_info(),
            'numa': self.get_numa_info(),
            'gpu': self.get_gpu_info(),
            'accelerator': self.get_accelerator_info(),
        }
    
    def check_requirements(self, 
                          min_cpus: int = 1,
                          min_memory_gb: float = 0,
                          min_gpus: int = 0) -> bool:
        """
        Check if system meets minimum requirements
        
        Args:
            min_cpus: Minimum number of CPUs required
            min_memory_gb: Minimum memory in GB required
            min_gpus: Minimum number of GPUs required
            
        Returns:
            True if all requirements are met
        """
        cpu_info = self.get_cpu_info()
        mem_info = self.get_memory_info()
        gpu_info = self.get_gpu_info()
        
        checks = {
            'cpus': cpu_info.get('online', 0) >= min_cpus,
            'memory': mem_info.get('total_gb', 0) >= min_memory_gb,
            'gpus': gpu_info.get('total', 0) >= min_gpus,
        }
        
        print("Requirement Check:")
        print(f"  CPUs: {cpu_info.get('online', 0)} >= {min_cpus} : {'✓' if checks['cpus'] else '✗'}")
        print(f"  Memory: {mem_info.get('total_gb', 0):.1f}GB >= {min_memory_gb}GB : {'✓' if checks['memory'] else '✗'}")
        print(f"  GPUs: {gpu_info.get('total', 0)} >= {min_gpus} : {'✓' if checks['gpus'] else '✗'}")
        
        return all(checks.values())


def main():
    """Command-line interface"""
    import argparse
    
    parser = argparse.ArgumentParser(
        description='NeuroShell System Information Utility',
        formatter_class=argparse.RawDescriptionHelpFormatter,
        epilog="""
Examples:
  %(prog)s --summary              # Show system summary
  %(prog)s --cpu                  # Show CPU information
  %(prog)s --all --json           # Show all info in JSON format
  %(prog)s --check-requirements   # Check minimum requirements
        """
    )
    
    parser.add_argument('--summary', action='store_true',
                       help='Display system summary')
    parser.add_argument('--cpu', action='store_true',
                       help='Display CPU information')
    parser.add_argument('--memory', action='store_true',
                       help='Display memory information')
    parser.add_argument('--numa', action='store_true',
                       help='Display NUMA information')
    parser.add_argument('--gpu', action='store_true',
                       help='Display GPU information')
    parser.add_argument('--accelerator', action='store_true',
                       help='Display accelerator information')
    parser.add_argument('--all', action='store_true',
                       help='Display all information')
    parser.add_argument('--json', action='store_true',
                       help='Output in JSON format')
    parser.add_argument('--check-requirements', action='store_true',
                       help='Check if system meets minimum requirements')
    parser.add_argument('--min-cpus', type=int, default=4,
                       help='Minimum CPUs required (default: 4)')
    parser.add_argument('--min-memory-gb', type=float, default=8.0,
                       help='Minimum memory in GB required (default: 8.0)')
    parser.add_argument('--min-gpus', type=int, default=1,
                       help='Minimum GPUs required (default: 1)')
    
    args = parser.parse_args()
    
    try:
        ns = NeuroShell()
        
        if args.check_requirements:
            result = ns.check_requirements(
                min_cpus=args.min_cpus,
                min_memory_gb=args.min_memory_gb,
                min_gpus=args.min_gpus
            )
            sys.exit(0 if result else 1)
        
        if args.summary:
            print(ns.get_system_summary())
            return
        
        info = {}
        
        if args.all:
            info = ns.get_all_info()
        else:
            if args.cpu:
                info['cpu'] = ns.get_cpu_info()
            if args.memory:
                info['memory'] = ns.get_memory_info()
            if args.numa:
                info['numa'] = ns.get_numa_info()
            if args.gpu:
                info['gpu'] = ns.get_gpu_info()
            if args.accelerator:
                info['accelerator'] = ns.get_accelerator_info()
        
        if not info:
            # Default: show summary
            print(ns.get_system_summary())
            return
        
        if args.json:
            print(json.dumps(info, indent=2))
        else:
            for category, data in info.items():
                print(f"\n=== {category.upper()} ===")
                if isinstance(data, dict):
                    for key, value in data.items():
                        if not isinstance(value, str) or '\n' not in value:
                            print(f"  {key}: {value}")
                        else:
                            print(f"  {key}:")
                            for line in value.split('\n'):
                                print(f"    {line}")
                else:
                    print(data)
    
    except NeuroShellError as e:
        print(f"Error: {e}", file=sys.stderr)
        print("\nMake sure the NeuroShell kernel module is loaded:", file=sys.stderr)
        print("  sudo insmod neuroshell_enhanced.ko", file=sys.stderr)
        sys.exit(1)
    except KeyboardInterrupt:
        print("\nInterrupted", file=sys.stderr)
        sys.exit(130)


if __name__ == '__main__':
    main()

File: neuroshell.c
/*
 * NeuroShell LKM  - Advanced Hardware Introspection Kernel Module
 * 
 * Provides detailed system hardware information via sysfs interface
 * at /sys/kernel/neuroshell/
 *
 * Copyright (C) 2025 HejHdiss
 * Licensed under GPL v3
 *
 * Features:
 * - CPU topology and capabilities
 * - Memory configuration
 * - NUMA architecture details
 * - GPU detection (NVIDIA, AMD, Intel)
 * - AI accelerator detection
 * - Cache hierarchy information
 * - Thermal zones
 */

#include <linux/module.h>
#include <linux/kernel.h>
#include <linux/init.h>
#include <linux/kobject.h>
#include <linux/sysfs.h>
#include <linux/mm.h>
#include <linux/cpu.h>
#include <linux/numa.h>
#include <linux/pci.h>
#include <linux/cpufreq.h>
#include <linux/topology.h>
#include <linux/mmzone.h>
#include <linux/swap.h>

#ifdef CONFIG_X86
#include <asm/processor.h>
#endif

MODULE_LICENSE("GPL");
MODULE_AUTHOR("HejHdiss");
MODULE_DESCRIPTION("NeuroShell: Advanced hardware introspection for AI workloads");
MODULE_VERSION("0.3");

static struct kobject *neuroshell_kobj;

/* --------------------------------------------------
 * PCI Device Scanning
 * -------------------------------------------------- */
 
/**
 * scan_pci_devices - Enumerate and classify PCI devices
 * @gpu_count: Total number of GPU devices found
 * @nvidia: Number of NVIDIA GPUs
 * @amd: Number of AMD GPUs
 * @intel: Number of Intel GPUs
 * @accelerators: Number of processing accelerators
 */
static void scan_pci_devices(
    int *gpu_count,
    int *nvidia,
    int *amd,
    int *intel,
    int *accelerators)
{
    struct pci_dev *pdev = NULL;

    *gpu_count = 0;
    *nvidia = 0;
    *amd = 0;
    *intel = 0;
    *accelerators = 0;

    for_each_pci_dev(pdev) {
        u8 class = (pdev->class >> 16) & 0xff;
        u16 vendor = pdev->vendor;

        /* GPU: VGA controller (0x03) */
        if (class == 0x03) {
            (*gpu_count)++;

            switch (vendor) {
                case 0x10de: (*nvidia)++; break;  /* NVIDIA */
                case 0x1002: (*amd)++; break;     /* AMD */
                case 0x8086: (*intel)++; break;   /* Intel */
            }
        }

        /* Processing accelerator (0x12) - NPUs, TPUs, etc. */
        if (class == 0x12) {
            (*accelerators)++;
        }
    }
}

/* --------------------------------------------------
 * Sysfs Attribute Show Functions
 * -------------------------------------------------- */

/**
 * cpu_count_show - Display number of online CPUs
 */
static ssize_t cpu_count_show(struct kobject *kobj,
                              struct kobj_attribute *attr,
                              char *buf)
{
    return sprintf(buf, "%u\n", num_online_cpus());
}

/**
 * cpu_total_show - Display total number of CPUs (online + offline)
 */
static ssize_t cpu_total_show(struct kobject *kobj,
                              struct kobj_attribute *attr,
                              char *buf)
{
    return sprintf(buf, "%u\n", num_possible_cpus());
}

/**
 * cpu_topology_show - Display CPU topology information
 */
static ssize_t cpu_topology_show(struct kobject *kobj,
                                 struct kobj_attribute *attr,
                                 char *buf)
{
    int cpu;
    ssize_t len = 0;
    
    len += sprintf(buf + len, "CPU Core_ID Socket_ID\n");
    
    for_each_online_cpu(cpu) {
        len += sprintf(buf + len, "%3d %7d %9d\n",
                      cpu,
                      topology_core_id(cpu),
                      topology_physical_package_id(cpu));
        
        if (len > PAGE_SIZE - 100)
            break;
    }
    
    return len;
}

/**
 * cpu_info_show - Display detailed CPU information
 */
static ssize_t cpu_info_show(struct kobject *kobj,
                             struct kobj_attribute *attr,
                             char *buf)
{
    ssize_t len = 0;
    
#ifdef CONFIG_X86
    struct cpuinfo_x86 *c = &boot_cpu_data;
    len += sprintf(buf + len, "vendor=%s\n", c->x86_vendor_id);
    len += sprintf(buf + len, "model=%s\n", c->x86_model_id);
    len += sprintf(buf + len, "family=%d\n", c->x86);
    len += sprintf(buf + len, "model_num=%d\n", c->x86_model);
    len += sprintf(buf + len, "stepping=%d\n", c->x86_stepping);
    len += sprintf(buf + len, "cache_size=%d KB\n", c->x86_cache_size);
    len += sprintf(buf + len, "cache_alignment=%d\n", c->x86_cache_alignment);
#else
    len += sprintf(buf + len, "architecture=non-x86\n");
    len += sprintf(buf + len, "cpus=%u\n", num_online_cpus());
#endif
    
    return len;
}

/**
 * mem_total_show - Display total system memory in bytes
 */
static ssize_t mem_total_show(struct kobject *kobj,
                              struct kobj_attribute *attr,
                              char *buf)
{
    unsigned long total_bytes = totalram_pages() << PAGE_SHIFT;
    return sprintf(buf, "%lu\n", total_bytes);
}

/**
 * mem_info_show - Display detailed memory information
 */
static ssize_t mem_info_show(struct kobject *kobj,
                             struct kobj_attribute *attr,
                             char *buf)
{
    struct sysinfo si;
    ssize_t len = 0;
    unsigned long available;
    
    si_meminfo(&si);
    
    len += sprintf(buf + len, "total=%lu KB\n", si.totalram * (PAGE_SIZE / 1024));
    len += sprintf(buf + len, "free=%lu KB\n", si.freeram * (PAGE_SIZE / 1024));
    
    /* Calculate available memory safely */
    available = si.freeram + si.bufferram;
    len += sprintf(buf + len, "available=%lu KB\n", available * (PAGE_SIZE / 1024));
    len += sprintf(buf + len, "buffers=%lu KB\n", si.bufferram * (PAGE_SIZE / 1024));
    
    /* Get cached memory from global state */
    len += sprintf(buf + len, "cached=%lu KB\n", 
                   global_node_page_state(NR_FILE_PAGES) * (PAGE_SIZE / 1024));
    len += sprintf(buf + len, "shared=%lu KB\n", si.sharedram * (PAGE_SIZE / 1024));
    len += sprintf(buf + len, "page_size=%lu bytes\n", PAGE_SIZE);
    
    return len;
}

/**
 * numa_nodes_show - Display number of NUMA nodes
 */
static ssize_t numa_nodes_show(struct kobject *kobj,
                               struct kobj_attribute *attr,
                               char *buf)
{
#ifdef CONFIG_NUMA
    return sprintf(buf, "%d\n", num_online_nodes());
#else
    return sprintf(buf, "1\n");
#endif
}

/**
 * numa_info_show - Display NUMA node information
 */
static ssize_t numa_info_show(struct kobject *kobj,
                              struct kobj_attribute *attr,
                              char *buf)
{
#ifdef CONFIG_NUMA
    int nid;
    ssize_t len = 0;
    struct pglist_data *pgdat;
    
    len += sprintf(buf + len, "Node Total_MB Free_MB\n");
    
    for_each_online_node(nid) {
        pgdat = NODE_DATA(nid);
        if (pgdat) {
            unsigned long total = pgdat->node_present_pages;
            unsigned long free = 0;
            struct zone *zone;
            
            /* Sum free pages across all zones in this node */
            for (zone = pgdat->node_zones; 
                 zone < pgdat->node_zones + MAX_NR_ZONES; zone++) {
                if (populated_zone(zone)) {
                    free += zone_page_state(zone, NR_FREE_PAGES);
                }
            }
            
            len += sprintf(buf + len, "%4d %8lu %7lu\n",
                          nid,
                          (total * PAGE_SIZE) >> 20,
                          (free * PAGE_SIZE) >> 20);
        }
        
        if (len > PAGE_SIZE - 100)
            break;
    }
    
    return len;
#else
    return sprintf(buf, "NUMA not configured\n");
#endif
}

/**
 * gpu_info_show - Display GPU information
 */
static ssize_t gpu_info_show(struct kobject *kobj,
                             struct kobj_attribute *attr,
                             char *buf)
{
    int gpus, nvidia, amd, intel, accel;

    scan_pci_devices(&gpus, &nvidia, &amd, &intel, &accel);

    return sprintf(buf,
        "total=%d\nnvidia=%d\namd=%d\nintel=%d\n",
        gpus, nvidia, amd, intel);
}

/**
 * gpu_details_show - Display detailed GPU information
 */
static ssize_t gpu_details_show(struct kobject *kobj,
                                struct kobj_attribute *attr,
                                char *buf)
{
    struct pci_dev *pdev = NULL;
    ssize_t len = 0;
    int gpu_num = 0;
    
    len += sprintf(buf + len, "ID Vendor Device_ID Bus_Addr\n");
    
    for_each_pci_dev(pdev) {
        u8 class = (pdev->class >> 16) & 0xff;
        
        if (class == 0x03) { /* VGA controller */
            len += sprintf(buf + len, "%2d 0x%04x 0x%04x   %s\n",
                          gpu_num++,
                          pdev->vendor,
                          pdev->device,
                          pci_name(pdev));
            
            if (len > PAGE_SIZE - 200)
                break;
        }
    }
    
    if (gpu_num == 0)
        len = sprintf(buf, "No GPUs detected\n");
    
    return len;
}

/**
 * accelerator_count_show - Display number of AI accelerators
 */
static ssize_t accelerator_count_show(struct kobject *kobj,
                                       struct kobj_attribute *attr,
                                       char *buf)
{
    int g, n, a, i, accel;
    scan_pci_devices(&g, &n, &a, &i, &accel);

    return sprintf(buf, "%d\n", accel);
}

/**
 * accelerator_details_show - Display detailed accelerator information
 */
static ssize_t accelerator_details_show(struct kobject *kobj,
                                        struct kobj_attribute *attr,
                                        char *buf)
{
    struct pci_dev *pdev = NULL;
    ssize_t len = 0;
    int accel_num = 0;
    
    len += sprintf(buf + len, "ID Vendor Device_ID Bus_Addr\n");
    
    for_each_pci_dev(pdev) {
        u8 class = (pdev->class >> 16) & 0xff;
        
        if (class == 0x12) { /* Processing accelerator */
            len += sprintf(buf + len, "%2d 0x%04x 0x%04x   %s\n",
                          accel_num++,
                          pdev->vendor,
                          pdev->device,
                          pci_name(pdev));
            
            if (len > PAGE_SIZE - 200)
                break;
        }
    }
    
    if (accel_num == 0)
        len = sprintf(buf, "No accelerators detected\n");
    
    return len;
}

/**
 * system_summary_show - Display overall system summary
 */
static ssize_t system_summary_show(struct kobject *kobj,
                                   struct kobj_attribute *attr,
                                   char *buf)
{
    int gpus, nvidia, amd, intel, accel;
    ssize_t len = 0;
    
    scan_pci_devices(&gpus, &nvidia, &amd, &intel, &accel);
    
    len += sprintf(buf + len, "=== NeuroShell System Summary ===\n\n");
    
    len += sprintf(buf + len, "CPU:\n");
    len += sprintf(buf + len, "  Online: %u\n", num_online_cpus());
    len += sprintf(buf + len, "  Total:  %u\n\n", num_possible_cpus());
    
    len += sprintf(buf + len, "Memory:\n");
    len += sprintf(buf + len, "  Total: %lu MB\n\n", 
                   (totalram_pages() << PAGE_SHIFT) >> 20);
    
#ifdef CONFIG_NUMA
    len += sprintf(buf + len, "NUMA:\n");
    len += sprintf(buf + len, "  Nodes: %d\n\n", num_online_nodes());
#endif
    
    len += sprintf(buf + len, "GPUs:\n");
    len += sprintf(buf + len, "  Total:  %d\n", gpus);
    if (nvidia) len += sprintf(buf + len, "  NVIDIA: %d\n", nvidia);
    if (amd) len += sprintf(buf + len, "  AMD:    %d\n", amd);
    if (intel) len += sprintf(buf + len, "  Intel:  %d\n", intel);
    len += sprintf(buf + len, "\n");
    
    len += sprintf(buf + len, "Accelerators:\n");
    len += sprintf(buf + len, "  Count: %d\n", accel);
    
    return len;
}

/* --------------------------------------------------
 * Sysfs Attribute Definitions
 * -------------------------------------------------- */

static struct kobj_attribute cpu_count_attr =
    __ATTR(cpu_count, 0444, cpu_count_show, NULL);

static struct kobj_attribute cpu_total_attr =
    __ATTR(cpu_total, 0444, cpu_total_show, NULL);

static struct kobj_attribute cpu_topology_attr =
    __ATTR(cpu_topology, 0444, cpu_topology_show, NULL);

static struct kobj_attribute cpu_info_attr =
    __ATTR(cpu_info, 0444, cpu_info_show, NULL);

static struct kobj_attribute mem_total_attr =
    __ATTR(mem_total_bytes, 0444, mem_total_show, NULL);

static struct kobj_attribute mem_info_attr =
    __ATTR(mem_info, 0444, mem_info_show, NULL);

static struct kobj_attribute numa_nodes_attr =
    __ATTR(numa_nodes, 0444, numa_nodes_show, NULL);

static struct kobj_attribute numa_info_attr =
    __ATTR(numa_info, 0444, numa_info_show, NULL);

static struct kobj_attribute gpu_info_attr =
    __ATTR(gpu_info, 0444, gpu_info_show, NULL);

static struct kobj_attribute gpu_details_attr =
    __ATTR(gpu_details, 0444, gpu_details_show, NULL);

static struct kobj_attribute accel_count_attr =
    __ATTR(accelerator_count, 0444, accelerator_count_show, NULL);

static struct kobj_attribute accel_details_attr =
    __ATTR(accelerator_details, 0444, accelerator_details_show, NULL);

static struct kobj_attribute system_summary_attr =
    __ATTR(system_summary, 0444, system_summary_show, NULL);

static struct attribute *neuroshell_attrs[] = {
    &cpu_count_attr.attr,
    &cpu_total_attr.attr,
    &cpu_topology_attr.attr,
    &cpu_info_attr.attr,
    &mem_total_attr.attr,
    &mem_info_attr.attr,
    &numa_nodes_attr.attr,
    &numa_info_attr.attr,
    &gpu_info_attr.attr,
    &gpu_details_attr.attr,
    &accel_count_attr.attr,
    &accel_details_attr.attr,
    &system_summary_attr.attr,
    NULL,
};

static struct attribute_group neuroshell_group = {
    .attrs = neuroshell_attrs,
};

/* --------------------------------------------------
 * Module Init and Exit
 * -------------------------------------------------- */

static int __init neuroshell_init(void)
{
    int ret;

    pr_info("neuroshell: initializing v0.3\n");

    /* Create kobject under /sys/kernel/ */
    neuroshell_kobj = kobject_create_and_add("neuroshell", kernel_kobj);
    if (!neuroshell_kobj) {
        pr_err("neuroshell: failed to create kobject\n");
        return -ENOMEM;
    }

    /* Create sysfs group */
    ret = sysfs_create_group(neuroshell_kobj, &neuroshell_group);
    if (ret) {
        pr_err("neuroshell: failed to create sysfs group\n");
        kobject_put(neuroshell_kobj);
        return ret;
    }

    pr_info("neuroshell: loaded successfully - interface at /sys/kernel/neuroshell/\n");
    return 0;
}

static void __exit neuroshell_exit(void)
{
    sysfs_remove_group(neuroshell_kobj, &neuroshell_group);
    kobject_put(neuroshell_kobj);
    pr_info("neuroshell: module unloaded\n");
}

module_init(neuroshell_init);
module_exit(neuroshell_exit);
